presto-main/src/main/java/com/facebook/presto/operator/unnest/UnnestOperatorBlockUtil.java

```diff
@@ -13,32 +13,304 @@
+import static com.facebook.presto.spi.block.ArrayBlock.fromElementBlock;
+import static com.facebook.presto.spi.block.MapBlock.fromKeyValueBlock;
+import static com.facebook.presto.spi.block.MethodHandleUtil.compose;
+import static com.facebook.presto.spi.block.MethodHandleUtil.nativeValueGetter;
+import static com.facebook.presto.spi.block.RowBlock.fromFieldBlocks;
+import static com.facebook.presto.testing.TestingEnvironment.TYPE_MANAGER;
 
 final class UnnestOperatorBlockUtil
 {
     private UnnestOperatorBlockUtil() {}
 
-    private static final int DEFAULT_CAPACITY = 64;
-    // See java.util.ArrayList for an explanation
-    static final int MAX_ARRAY_SIZE = Integer.MAX_VALUE - 8;
+    static Block insertNull(Block block, Type type)
+    {
+        if (block instanceof DictionaryBlock) {
+            return insertNullToDictionaryBlock((DictionaryBlock) block, type);
+        }
+
+        if (block instanceof RunLengthEncodedBlock) {
+            return insertNullToRleBlock((RunLengthEncodedBlock) block, type);
+        }
+
+        if (block instanceof ByteArrayBlock) {
+            return insertNullToByteArrayBlock((ByteArrayBlock) block);
+        }
+
+        if (block instanceof ShortArrayBlock) {
+            return insertNullToShortArrayBlock((ShortArrayBlock) block);
+        }
+
+        if (block instanceof IntArrayBlock) {
+            return insertNullToIntArrayBlock((IntArrayBlock) block);
+        }
+
+        if (block instanceof LongArrayBlock) {
+            return insertNullToLongArrayBlock((LongArrayBlock) block);
+        }
+
+        if (block instanceof Int128ArrayBlock) {
+            return insertNullToInt128ArrayBlock((Int128ArrayBlock) block);
+        }
+
+        if (block instanceof VariableWidthBlock) {
+            return insertNullToVariableWidthBlock((VariableWidthBlock) block);
+        }
+
+        if (block instanceof ArrayBlock) {
+            return insertNullToArrayBlock((ArrayBlock) block);
+        }
+
+        if (block instanceof MapBlock) {
+            return insertNullToMapBlock((MapBlock) block, (MapType) type);
+        }
+
+        if (block instanceof RowBlock) {
+            return insertNullToRowBlock((RowBlock) block);
+        }
+
+        throw new PrestoException(NOT_SUPPORTED, "Cannot insert null to " + block.getClass());
+    }
 
-    // Copied from io.prestosql.spi.block.BlockUtil#calculateNewArraySize
-    static int calculateNewArraySize(int currentSize)
+    private static Block insertNullToDictionaryBlock(DictionaryBlock block, Type type)
     {
-        // grow array by 50%
-        long newSize = (long) currentSize + (currentSize >> 1);
+        int positionCount = block.getPositionCount();
+        Block dictionary = block.getDictionary();
 
-        // verify new size is within reasonable bounds
-        if (newSize < DEFAULT_CAPACITY) {
-            newSize = DEFAULT_CAPACITY;
+        int[] ids = new int[positionCount + 1];
```

Comment:
perhaps, use `getIds`: `int[] ids = Arrays.copyOf(block.getRawIds(), positionCount + 1);`

Reply:
Hi Masha, getRawIds() is package private and can't be accessed by the operator. After we add appendNull() to Block interface these won't be needed.